File: 2021/python/Day-02/Day_02.py
import time

with open("input.txt","r") as f:
    input_list = [(action, int(v)) for action, v in (i.strip().split() for i in f.readlines())]


def solution01(input_list):
    depth, pos_horizontal = 0,0

    for action, v in input_list:

        if action == "forward":
            pos_horizontal += v
        elif action == "down":
            depth += v
        elif action == "up":
            depth -= v

    return depth * pos_horizontal

def solution02(input_list):
    depth, pos_horizontal, aim = 0,0,0

    for action, v in input_list:

        if action == "forward":
            pos_horizontal += v
            depth += aim * v
        elif action == "down":
            aim += v
        elif action == "up":
            aim -= v

    return depth * pos_horizontal


solution01_start_time = time.time()
solution01 = solution01(input_list)
solution01_total_time = time.time() - solution01_start_time
print(f"Solution 01: {solution01}, in {solution01_total_time}")

solution02_start_time = time.time()
solution02 = solution02(input_list)
solution02_total_time = time.time() - solution02_start_time
print(f"Solution 02: {solution02}, in {solution02_total_time}")

File: 2021/js/Day-02/Day-02.js
const fs = require("fs");

inputData = fs.readFileSync("input.txt", "utf8");

inputData = inputData
  .split(/\r?\n/)
  .map((v) => v.split(" "))
  .map((v) => [v[0], Number(v[1])]);

const Solution01 = (inputData) => {
  let horizontal = 0,
    depth = 0;

  inputData.forEach((el) => {
    let [instruction, v] = el;

    switch (instruction) {
      case "forward":
        horizontal += v;
        break;
      case "up":
        depth -= v;
        break;
      case "down":
        depth += v;
        break;
    }
  });

  return depth * horizontal;
};

const Solution02 = (inputData) => {
  let horizontal = 0,
    depth = 0,
    aim = 0;

  inputData.forEach((el) => {
    let [instruction, v] = el;

    switch (instruction) {
      case "forward":
        horizontal += v;
        depth += aim * v;
        break;
      case "up":
        aim -= v;
        break;
      case "down":
        aim += v;
        break;
    }
  });

  return depth * horizontal;
};

console.log(`Solution 01: ${Solution01(inputData)}`);
console.log(`Solution 02: ${Solution02(inputData)}`);
